### PR TITLE
HLSL compiler updates

### PR DESCRIPTION
**Pull request by deprilula28**
## Description
- Adds support for custom spirv optimizer settings
- Returns the spirv shader directly on preprocess shader

## Testing 

- Compilation was tested with and without a custom optimizer 

## TODO list:

- [x] Figure out how to best get the latest shader model supported
- [x] Correct line reporting by DXC
- [x] Log warnings when there is no error

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
Also remember about: https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1116473899

**Comment by devshgraphicsprogramming** [comment]
Also to quote Erfan
```
I'm getting incorrect error/warning line and file

Add these TODOs to the hlsl fixes PR:

- correct line reporting by dxc (#line) you can test my cad app and mess with the shaders

- add an option to log warnings as well, i find some of dxc warnings important but i only get them if there is a compile error which prints everything dxc is reporting
```

File: src/nbl/asset/utils/CHLSLCompiler.cpp [hunk 1]
```diff
@@ -57,7 +59,10 @@ static tcpp::IInputStream* getInputStreamInclude(
     uint32_t _maxInclCnt,
     const char* _requesting_source,
     const char* _requested_source,
-    bool _type // true for #include "string"; false for #include <string>
+    bool _type, // true for #include "string"; false for #include <string>
```
Comment (Erfan-Ahmadi) [inline src/nbl/asset/utils/CHLSLCompiler.cpp:62]: name this "isRelative" also remove the weird `_` before var names and make them consistent (camelCase)

see how shaderc names them:
```c++
// The kinds of include requests.
enum shaderc_include_type {
  shaderc_include_type_relative,  // E.g. #include "source"
  shaderc_include_type_standard   // E.g. #include <source>
};
```

File: src/nbl/asset/utils/CHLSLCompiler.cpp [hunk 2]
```diff
@@ -169,13 +199,24 @@ DxcCompilationResult dxcCompile(const CHLSLCompiler* compiler, nbl::asset::hlsl:
 
 std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
 {
+    std::ostringstream insertion;
+    insertion << IShaderCompiler::PREPROC_DIRECTIVE_ENABLER;
+    insertion << "line 1\n";
+    insertIntoStart(code, std::move(insertion));
```
Comment (Erfan-Ahmadi) [inline src/nbl/asset/utils/CHLSLCompiler.cpp:205]: just use [std::string::insert](https://cplusplus.com/reference/string/string/insert/) in a single liner

File: src/nbl/asset/utils/IShaderCompiler.cpp [hunk 3]
```diff
@@ -70,14 +66,25 @@ std::string IShaderCompiler::encloseWithinExtraInclGuards(std::string&& _code, u
...
         "#endif"
         "\n\n" +
         genUndefs();
 }
 
+// Amount of lines before the #line after having run encloseWithinExtraInclGuards
+uint32_t IShaderCompiler::encloseWithinExtraInclGuardsLeadingLines(uint32_t _maxInclusions)
+{
+    auto lineDirectiveString = std::string(IShaderCompiler::PREPROC_DIRECTIVE_DISABLER) + "line";
+    std::string str = IShaderCompiler::encloseWithinExtraInclGuards(std::string(""), _maxInclusions, "encloseWithinExtraInclGuardsLeadingLines");
```
Comment (Erfan-Ahmadi) [inline src/nbl/asset/utils/IShaderCompiler.cpp:81]: shouldn't we use the name of the file here instead? or something?

File: src/nbl/asset/utils/CHLSLCompiler.cpp [hunk 4]
```diff
@@ -87,8 +92,25 @@ static tcpp::IInputStream* getInputStreamInclude(
...
+        (includeStack.size() > 1 ? leadingLinesImports : 0);
+
     IShaderCompiler::disableAllDirectivesExceptIncludes(res_str);
     res_str = IShaderCompiler::encloseWithinExtraInclGuards(std::move(res_str), _maxInclCnt, name.string().c_str());
+    res_str = res_str + "\n" +
+        IShaderCompiler::PREPROC_DIRECTIVE_DISABLER + "line " + std::to_string(lineGoBackTo) + " \"" +  includeStack.back().second + "\"\n";
+
+    // HACK: tcpp is having issues parsing the string, so this is a hack/mitigation that could be removed once tcpp is fixed
+    std::string identifier = name.string().c_str();
+    std::replace(identifier.begin(), identifier.end(), '\\', '/');
```
Comment (Erfan-Ahmadi) [inline src/nbl/asset/utils/CHLSLCompiler.cpp:109]: Do we still need this fix after the tcpp guy fixed that?

File: src/nbl/asset/utils/IShaderCompiler.cpp (hunk 3 above)
Comment (Erfan-Ahmadi) [reply src/nbl/asset/utils/IShaderCompiler.cpp:81]: I guess it's fine, you're just counting the lines :D

File: src/nbl/asset/utils/CHLSLCompiler.cpp [hunk 5]
```diff
@@ -239,6 +284,8 @@ std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADE
 
     auto resolvedString = proc.Process();
     IShaderCompiler::reenableDirectives(resolvedString);
+
+    printf("Resolved string:\n\n%s\n", resolvedString.c_str());
```
Comment (Erfan-Ahmadi) [inline src/nbl/asset/utils/CHLSLCompiler.cpp:288]: I think this is for debugging?

Comment (deprilula28) [reply src/nbl/asset/utils/CHLSLCompiler.cpp:288]: oops

File: src/nbl/asset/utils/IShaderCompiler.cpp [hunk 6]
```diff
@@ -12,6 +12,19 @@
...
 
+std::string IShaderCompiler::escapeFilename(std::string&& code)
+{
+    std::string dest;
+    for (char c : code)
+    {
+        if (c == '\\')
+            dest.append("\\" "\\");
+        else 
+            dest.push_back(c);
```
[the remark this answers is not in the document]
Comment (deprilula28) [reply src/nbl/asset/utils/IShaderCompiler.cpp:23]: This is all that's needed, you just need to replace the \ with \\

[the remark this answers is not in the document]
Comment (deprilula28) [reply src/nbl/asset/utils/IShaderCompiler.cpp:23]: There aren't any unescaped \n s within the string

**Comment by devshgraphicsprogramming** [comment]
@Erfan-Ahmadi can we merge?

**Comment by Erfan-Ahmadi** [comment]
> @Erfan-Ahmadi can we merge?

I still haven't reviewed the part with fixing line includes. It works though.

**Comment by devshgraphicsprogramming** [comment]
@deprilula28 is there any issue in yanking up the SM from 6.2 to 6.6 ?
https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_Pack_Unpack_Intrinsics.html

File: src/nbl/asset/utils/CHLSLCompiler.cpp [hunk 7]
```diff
@@ -169,13 +197,22 @@ DxcCompilationResult dxcCompile(const CHLSLCompiler* compiler, nbl::asset::hlsl:
 
 std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
 {
+    // Line 1 comes before all the extra defines in the main shader
+    insertIntoStart(code, std::ostringstream(std::string(IShaderCompiler::PREPROC_DIRECTIVE_ENABLER) + "line 1\n"));
+
+    uint32_t defineLeadingLinesMain = 0;
```
Comment (Erfan-Ahmadi) [inline src/nbl/asset/utils/CHLSLCompiler.cpp:203]: I think this need to be 1?

**Pull request merged**